source/opt/ir_context.h

```diff
@@ -121,8 +132,64 @@ class IRContext {
   // Appends a function to this module.
   inline void AddFunction(std::unique_ptr<Function>&& f);
 
+  // Returns a pointer to a def-use manager.  If the def-use manager is
+  // invalid, it is rebuilt first.
+  opt::analysis::DefUseManager* get_def_use_mgr() {
+    if (!AreAnalysesValid(kAnalysisDefUse)) BuildDefUseManager();
+    return def_use_mgr_.get();
+  }
+
+  // Builds the def-use manager from scratch, even if it was already valid.
+  void BuildDefUseManager() {
+    def_use_mgr_.reset(new opt::analysis::DefUseManager(module()));
+    valid_analyses_ = (Analysis)(valid_analyses_ | kAnalysisDefUse);
+  }
+
+  // Sets the message consumer to the given |consumer|. |consumer| which will be
+  // invoked every time there is a message to be communicated to the outside.
+  void SetMessageConsumer(spvtools::MessageConsumer c) {
+    consumer_ = std::move(c);
+  }
+
+  // Returns the reference to the message consumer for this pass.
+  const spvtools::MessageConsumer& consumer() const { return consumer_; }
+
+  // Rebuilds the analyses in |set| that are invalid.
+  void BuildAnalyses(Analysis set);
```

Comment:
Slight semantic confusion.  BuildDefUseManager() is a forced build, but BuildAnalyses() is not.  Perhaps rename BuildAnalyses() to BuildInvalidAnalyses()?

Reply:
Or GetAnalyses, with method comment that it will reuse valid analyses, and rebuild invalid analysis.

Reply:
I went with BuildInvalidAnalyses.  The "GetAnalyes" might make people think something is being returned, when it is not.